File: java/109.java



// Runtime: 200 ms, faster than 5.80% of Java online submissions for Convert Sorted List to Binary Search Tree.
// Memory Usage: 44.5 MB, less than 5.26% of Java online submissions for Convert Sorted List to Binary Search Tree.

class Solution {
  public TreeNode sortedListToBST(ListNode head) {
    if (Objects.isNull(head)) {
     return null;
    }
    int length = getLength(head);
      return subtree(head, 0, length - 1);
    }

    private TreeNode subtree(ListNode head, int start, int end) {
      TreeNode root = null;
      if (start <= end) {
        int mid = (start + end) % 2 == 0 ? (start + end) / 2
               : (start + end) / 2 + 1;
        root = new TreeNode(getNode(head, mid).val);
        root.left = subtree(head, start, mid - 1);
        root.right = subtree(head, mid + 1, end);
      }
      return root;
    }

    private ListNode getNode(ListNode head, int index) {
      for (int i = 0; i < index; i++) {
        head = head.next;
      }
      return head;
    }

    private int getLength(ListNode head) {
      int count = 0;
      while (Objects.nonNull(head)) {
        head = head.next;
        count++;
      }
      return count;
    }
}


// Runtime: 2 ms, faster than 9.48% of Java online submissions for Convert Sorted List to Binary Search Tree.
// Memory Usage: 44.7 MB, less than 5.26% of Java online submissions for Convert Sorted List to Binary Search Tree.
class Solution {
  public TreeNode sortedListToBST(ListNode head) {
    if (Objects.isNull(head)) {
      return null;
    }
    List<Integer> list = new ArrayList<>();
    ListNode cur = head;
    while (Objects.nonNull(cur)) {
      list.add(cur.val);
      cur = cur.next;
    }
    return subtree(list, 0, list.size() - 1);
  }

  private TreeNode subtree(List list, int start, int end) {
    TreeNode root = null;
    if (start <= end) {
      int mid = (start + end) % 2 == 0 ? (start + end) / 2
              : (start + end) / 2 + 1;
      root = new TreeNode((int)list.get(mid));
      root.left = subtree(list, start, mid - 1);
      root.right = subtree(list, mid + 1, end);
    }
    return root;
  }
}

File: go/109.go
// 109. Convert Sorted List to Binary Search Tree

// Given the head of a singly linked list where elements are sorted in ascending order, convert it to a height-balanced binary search tree.

 

// Example 1:


// Input: head = [-10,-3,0,5,9]
// Output: [0,-3,9,-10,null,5]
// Explanation: One possible answer is [0,-3,9,-10,null,5], which represents the shown height balanced BST.
// Example 2:

// Input: head = []
// Output: []
 

// Constraints:

// The number of nodes in head is in the range [0, 2 * 104].
// -105 <= Node.val <= 105

// Runtime 0 ms Beats 100.00%
// Memory 7.89 MB Beats 70.51%
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sortedListToBST(head *ListNode) *TreeNode {
    if head == nil {
        return nil
    }
    if head.Next == nil {
        return &TreeNode{Val: head.Val}
    }
    mid := getMid(head)
    root := &TreeNode{Val: mid.Val}
    left := sortedListToBST(head)
    right := sortedListToBST(mid.Next)
    root.Left = left
    root.Right = right
    return root
}

func getMid(head *ListNode) *ListNode {
    fast := head
    slow := head
    var prev *ListNode
    prev = nil
    for ; fast != nil && fast.Next != nil; {
        prev = slow
        slow = slow.Next
        fast = fast.Next.Next
    }
    prev.Next = nil
    return slow
}